xla/service/reduce_scatter_all_gather_combiner_test.cc

```diff
@@ -0,0 +1,161 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/reduce_scatter_all_gather_combiner.h"
+
+#include "tensorflow/compiler/xla/hlo/ir/hlo_casting_utils.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instructions.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_module.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_opcode.h"
+#include "tensorflow/compiler/xla/hlo/utils/hlo_matchers.h"
+#include "tensorflow/compiler/xla/service/hlo_parser.h"
+#include "tensorflow/compiler/xla/service/hlo_pass_pipeline.h"
+#include "tensorflow/compiler/xla/service/hlo_verifier.h"
+#include "tensorflow/compiler/xla/tests/hlo_test_base.h"
+#include "tensorflow/compiler/xla/util.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
+#include "tensorflow/tsl/lib/core/status_test_util.h"
+
+namespace xla {
+namespace {
+
+namespace op = xla::testing::opcode_matchers;
+
+class ReduceScatterAllGatherCombinerTest : public HloTestBase {
+ public:
+  StatusOr<std::unique_ptr<HloModule>> RunPass(absl::string_view hlo_module,
+                                               int64_t num_replicas,
+                                               int64_t num_partitions,
+                                               bool expect_change) {
+    HloModuleConfig config = GetModuleConfigForTest(
+        /*replica_count=*/num_replicas,
+        /*num_partitions=*/num_partitions);
+    config.set_use_spmd_partitioning(num_partitions > 1);
+    TF_ASSIGN_OR_RETURN(auto module,
+                        ParseAndReturnVerifiedModule(hlo_module, config));
+
+    auto changed = ReduceScatterAllGatherCombiner().Run(module.get());
+    if (!changed.ok()) {
+      return changed.status();
+    }
+    EXPECT_EQ(changed.value(), expect_change);
+    return StatusOr<std::unique_ptr<HloModule>>(std::move(module));
+  }
+
+  template <HloOpcode oc>
+  size_t CollectiveCount(std::unique_ptr<HloModule> &module) {
+    return absl::c_count_if(module->entry_computation()->instructions(),
+                            HloPredicateIsOp<oc>);
+  }
+};
+
+TEST_F(ReduceScatterAllGatherCombinerTest, BranchesOptimized) {
```

Comment:
It seems there are 2 independent patterns being combined and tested. I'd have expected a first simpler test that just tests a single pair of RS and AG combined, and call it something like SImpleCombine?

Also, why is the test called BranchesOptimized? I think it might confuse future readers. You can keep this test and call it MultiplePatterns or something but I think a basic test of just one RS+AR makes sense.

Reply:
added a new test and changed the name